### PR TITLE
Range bugfix

### PR DESCRIPTION
**Pull request by renehiemstra**
Refactored entire implementation. Iteration over a range is now implemented in terms of iterators, which have pointer semantics:
1. `var iter = range:getfirst()` -> yields the iterator at the start.
2. `range:getvalue(&iter)` -> yields the current value
3. `range:next(&iter)` -> increments the iterator.
4. `range:isvalid(&iter)` -> yields a boolean value that signals whether the iterator points to a valid value.

If you implement the above iterator interface then you can use the range object in a for loop
```
for v in range do
   --run body(v)
end
```
which generates the following loop:
```
var iter = range:getfirst()
while range:isvalid(&iter) do
    var value = range:getvalue(&iter)
    [body(value)] --run body of loop
    range:next(&iter) --increment state
end
```

* It may be possible to combine `range:next` and `range:isvalid` into one function. We can investigate that later. 
* We may want to put a larger focus on iterators and do something like the following:
```
var iter = range:getiter()
```
where `iter` holds a reference to the range object and then implement an interface:
1. `iter:getvalue()` -> yields the current value
2. `iter:next()` -> increments the iterator.
3. `iter:isvalid()` -> yields a boolean value that signals whether the iterator points to a valid value.

### CODE REVIEW

File: example_stack_heap.t [hunk 1]
```diff
@@ -7,6 +7,7 @@ require("terralibext")
 local alloc = require("alloc")
 local err = require("assert")
 local range = require("range")
+local io = terralib.includec("stdio.h")
```
Comment (tpkessler) [inline example_stack_heap.t:10]: Can we remove this line? It doesn't seem to be necessary to include it.

File: range.t [hunk 2]
```diff
@@ -55,16 +55,24 @@ local RangeBase = function(Range, state_t, T)
         end
     end)
 
+    --always extract a value type into the body of the loop
+    local extract = function(value) 
+        if value.type:ispointer() then
```
Comment (tpkessler) [inline range.t:60]: What happens if `value.type` is `nil`?

File: example_stack_heap.t (hunk 1 above)
Comment (renehiemstra) [reply example_stack_heap.t:10]: yes

**Comment by tpkessler** [comment]
Do we still need this?

**Comment by renehiemstra** [comment]
yes, just keep it there for now. I've updated it as part of the alloc pull request

**Comment by renehiemstra** [comment]
I've refactored once more and centered the implementation of ranges around iterators that implement the iterator interface, as described on top.

**Pull request merged**